fluss-server/src/main/java/com/alibaba/fluss/server/replica/Replica.java

```diff
@@ -880,28 +888,34 @@ private boolean maybeIncrementLeaderHW(LogTablet leaderLog, long currentTimeMs)
         // maybeIncrementLeaderHW is in the hot path, the following code is written to
         // avoid unnecessary collection generation.
         LogOffsetMetadata leaderLogEndOffset = leaderLog.getLocalEndOffsetMetadata();
-        LogOffsetMetadata newHighWatermark = leaderLogEndOffset;
+        LogOffsetMetadata newAckedLogEndOffset = leaderLogEndOffset;
 
         for (FollowerReplica remoteFollowerReplica : followerReplicasMap.values()) {
             // Note here we are using the "maximal", see explanation above.
             FollowerReplica.FollowerReplicaState replicaState =
                     remoteFollowerReplica.stateSnapshot();
             int followerId = remoteFollowerReplica.getFollowerId();
             if (replicaState.getLogEndOffsetMetadata().getMessageOffset()
-                            < newHighWatermark.getMessageOffset()
+                            < newAckedLogEndOffset.getMessageOffset()
                     && (isrState.maximalIsr().contains(followerId)
                             || shouldWaitForReplicaToJoinIsr(
                                     replicaState, leaderLogEndOffset, currentTimeMs, followerId))) {
-                newHighWatermark = replicaState.getLogEndOffsetMetadata();
+                newAckedLogEndOffset = replicaState.getLogEndOffsetMetadata();
             }
         }
 
-        Optional<LogOffsetMetadata> oldWatermark =
-                leaderLog.maybeIncrementHighWatermark(newHighWatermark);
-        if (oldWatermark.isPresent()) {
-            LOG.debug("High watermark update from {} to {}.", oldWatermark.get(), newHighWatermark);
-            // when watermark advanced, we may need to flush kv if it's kv replica
-            mayFlushKv(newHighWatermark.getMessageOffset());
+        Optional<LogOffsetMetadata> optOldAckedLogEndOffset =
+                leaderLog.maybeIncrementAckedLogEndOffset(newAckedLogEndOffset);
+        if (optOldAckedLogEndOffset.isPresent()) {
+            long messageOffset = optOldAckedLogEndOffset.get().getMessageOffset();
+            LOG.debug(
+                    "AckedLogEndOffset update from {} to {}.", messageOffset, newAckedLogEndOffset);
+
+            // when ackedLogEndOffset advanced, we may need to flush kv if it's kv replica
+            mayFlushKv(newAckedLogEndOffset.getMessageOffset());
```

Comment:
maybe we can simplify flush kv first, and then update high-watermark? Our purpose is just make KV ready before log is available. This can avoid introducing additional variable `ackedLogEndOffsetMetadata` (it is quite complex if we need to handle the restore of the new variable).

Reply:
Done!